File: python/p0036.py
from typing import List
from collections import defaultdict

class Solution:
  def isValidSudoku(self, board: List[List[str]]) -> bool:
    # Initialize dictionaries to track seen numbers in rows, columns, and squares
    rows = defaultdict(set)
    columns = defaultdict(set)
    squares = defaultdict(set)

    # Iterate over each cell in the 9x9 Sudoku board
    for r in range(9):
      for c in range(9):
        # Skip empty cells
        if board[r][c] == ".":
          continue

        # Check if the number is already seen in the current row, column, or 3x3 square
        if board[r][c] in rows[r] or board[r][c] in columns[c] or board[r][c] in squares[(r // 3, c // 3)]:
          return False  # If duplicate is found, Sudoku is invalid

        # Add the number to the respective row, column, and square sets
        rows[r].add(board[r][c])
        columns[c].add(board[r][c])
        squares[(r // 3, c // 3)].add(board[r][c])

    return True  # If no duplicates are found, Sudoku is valid

File: javascript/p0036.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
export var isValidSudoku = function (board) {
  // Initialize dictionaries to track seen numbers in rows, columns, and squares
  const rows = Array.from({ length: 9 }, () => new Set());
  const columns = Array.from({ length: 9 }, () => new Set());
  const squares = Array.from({ length: 9 }, () => new Set());

  // Iterate over each cell in the 9x9 Sudoku board
  for (let r = 0; r < 9; r++) {
    for (let c = 0; c < 9; c++) {
      // Skip empty cells
      if (board[r][c] === ".") continue;

      const num = board[r][c];
      const squareIndex = Math.floor(r / 3) * 3 + Math.floor(c / 3);

      // Check if the number is already seen in the current row, column, or 3x3 square
      if (
        rows[r].has(num) ||
        columns[c].has(num) ||
        squares[squareIndex].has(num)
      ) {
        return false; // If duplicate is found, Sudoku is invalid
      }

      // Add the number to the respective row, column, and square sets
      rows[r].add(num);
      columns[c].add(num);
      squares[squareIndex].add(num);
    }
  }

  return true; // If no duplicates are found, Sudoku is valid
};
